File: frontend/src/api/bridge.ts
/**
 * 双向桥梁 API 封装
 * Lv.1 你好桥 / Lv.2 点赞桥 / Lv.3 梗桥
 */
const BRIDGE_BASE = '/api/v2/bridge'

interface ApiResponse<T> {
  status: string
  data: T
}

function getUserId(): string {
  try {
    const raw = localStorage.getItem('lawa2_session')
    if (!raw) return 'default_user'
    const s = JSON.parse(raw)
    return s?.userId || 'default_user'
  } catch {
    return 'default_user'
  }
}

async function bridgeFetch<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, {
    headers: { 'Content-Type': 'application/json' },
    ...init,
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail || `HTTP ${res.status}`)
  }
  const json: ApiResponse<T> = await res.json()
  return json.data
}

function withUser(path: string): string {
  const uid = getUserId()
  const sep = path.includes('?') ? '&' : '?'
  return `${BRIDGE_BASE}${path}${sep}user_id=${encodeURIComponent(uid)}`
}

// ── 数据模型 ──

export interface BridgePartner {
  partner_id: string
  partner_name: string
  native_lang: string
  learn_lang: string
  bio: string
  bridge_level: number
}

export interface BridgeGreeting {
  interaction_id: string
  partner_name: string
  greeting: string
  translation: string
  context: string
  direction: string
  created_at: string
}

export interface BridgeReply {
  status: string
  your_reply: string
  polished: string
  partner_reply: string
  xp_earned: number
}

export interface BridgeLikePrompt {
  interaction_id: string
  partner_name: string
  message: string
  translation: string
  context: string
  level: number
  direction: string
  created_at: string
}

export interface BridgeTeachPrompt {
  interaction_id: string
  partner_name: string
  message: string
  translation: string
  context: string
  level: number
  direction: string
  created_at: string
}

export interface BridgeTeachResult {
  status: string
  word: string
  meaning: string
  partner_curious: string
  partner_try_use: string
  partner_thanks: string
  xp_earned: number
}

export interface BridgeHistory {
  id: string
  level: number
  direction: string
  learn_text: string
  native_text: string
  polished_text: string | null
  partner_reply: string | null
  xp_earned: number
  created_at: string
}

// ── Lv.4 群聊桥 ──

export interface GroupMember {
  name: string
  role: string
}

export interface GroupMessage {
  from: string
  text: string
}

export interface BridgeGroupPrompt {
  interaction_id: string
  partner_name: string
  scene: string
  context: string
  members: GroupMember[]
  initial_messages: GroupMessage[]
  messages_text: string
  members_text: string
  level: number
  direction: string
  created_at: string
}

export interface BridgeGroupResult {
  status: string
  your_reply: string
  group_replies: GroupMessage[]
  replies_text: string
  xp_earned: number
}

// ── Lv.5 线下桥 ──

export interface OfflineNpc {
  name: string
  role: string
}

export interface BridgeOfflinePrompt {
  interaction_id: string
  partner_name: string
  scene: string
  context: string
  npc: OfflineNpc
  initial_message: string
  level: number
  direction: string
  created_at: string
}

export interface BridgeOfflineResult {
  status: string
  your_reply: string
  npc_reply: string
  npc_action: string
  culture_tip: string
  full_reply: string
  xp_earned: number
}

export interface BridgeProgress {
  current_level: number
  total_interactions: number
  total_xp: number
  next_level_at: string
  levels: Array<{
    level: number
    name: string
    unlocked: boolean
    done: boolean
  }>
}

// ── API 方法 ──

export async function getBridgePartner(): Promise<BridgePartner> {
  return bridgeFetch(withUser('/partner'))
}

export async function getBridgeGreeting(): Promise<BridgeGreeting> {
  return bridgeFetch(withUser('/greeting'))
}

export async function replyBridgeGreeting(interactionId: string, replyText: string): Promise<BridgeReply> {
  return bridgeFetch(withUser('/reply'), {
    method: 'POST',
    body: JSON.stringify({ interaction_id: interactionId, reply_text: replyText }),
  })
}

export async function getBridgeHistory(limit = 10): Promise<BridgeHistory[]> {
  return bridgeFetch(withUser(`/history?limit=${limit}`))
}

export async function getBridgeProgress(): Promise<BridgeProgress> {
  return bridgeFetch(withUser('/progress'))
}

// ── Lv.2 点赞桥 ──

export async function getBridgeLikePrompt(): Promise<BridgeLikePrompt> {
  return bridgeFetch(withUser('/like'))
}

export async function replyBridgeLike(interactionId: string, replyText: string): Promise<BridgeReply> {
  return bridgeFetch(withUser('/like'), {
    method: 'POST',
    body: JSON.stringify({ interaction_id: interactionId, reply_text: replyText }),
  })
}

// ── Lv.4 群聊桥 ──

export async function getBridgeGroupPrompt(): Promise<BridgeGroupPrompt> {
  return bridgeFetch(withUser('/group'))
}

export async function replyBridgeGroup(interactionId: string, replyText: string): Promise<BridgeGroupResult> {
  return bridgeFetch(withUser('/group'), {
    method: 'POST',
    body: JSON.stringify({ interaction_id: interactionId, reply_text: replyText }),
  })
}

// ── Lv.5 线下桥 ──

export async function getBridgeOfflinePrompt(): Promise<BridgeOfflinePrompt> {
  return bridgeFetch(withUser('/offline'))
}

export async function replyBridgeOffline(interactionId: string, replyText: string): Promise<BridgeOfflineResult> {
  return bridgeFetch(withUser('/offline'), {
    method: 'POST',
    body: JSON.stringify({ interaction_id: interactionId, reply_text: replyText }),
  })
}

// ── Lv.3 梗桥 ──

export async function getBridgeTeachPrompt(): Promise<BridgeTeachPrompt> {
  return bridgeFetch(withUser('/teach'))
}

export async function teachBridgeWord(
  interactionId: string, word: string, meaning: string, example = ''
): Promise<BridgeTeachResult> {
  return bridgeFetch(withUser('/teach'), {
    method: 'POST',
    body: JSON.stringify({ interaction_id: interactionId, word, meaning, example }),
  })
}

File: src/routes/bridge.py
"""
LAWA 双向桥梁 — API 路由

Lv.1 你好桥：
  - GET  /api/v2/bridge/partner     — 获取语伴信息
  - GET  /api/v2/bridge/greeting    — 获取问候语（模拟语伴发消息）
  - POST /api/v2/bridge/reply       — 回复问候语（AI 润色 + 反馈 + 奖励）
  - GET  /api/v2/bridge/history     — 交互历史
  - GET  /api/v2/bridge/progress    — 桥梁进度

Lv.2 点赞桥：
  - GET  /api/v2/bridge/like        — 获取点赞桥内容（语伴分享话题）
  - POST /api/v2/bridge/like        — 用户评价语伴分享的内容

Lv.3 梗桥：
  - GET  /api/v2/bridge/teach       — 获取教梗提示（语伴想学词）
  - POST /api/v2/bridge/teach       — 用户教语伴一个词
"""
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from loguru import logger

from src.engine.bridge_engine import BridgeEngine


router = APIRouter(prefix="/api/v2/bridge", tags=["bridge"])


# ── 依赖注入 ──

def get_bridge_engine() -> BridgeEngine:
    return BridgeEngine()


async def get_user_id(user_id: str = "default_user") -> str:
    return user_id


# ── 请求模型 ──

class ReplyRequest(BaseModel):
    interaction_id: str
    reply_text: str

class LikeRequest(BaseModel):
    interaction_id: str
    reply_text: str

class TeachRequest(BaseModel):
    interaction_id: str
    word: str
    meaning: str
    example: str = ""

class GroupReplyRequest(BaseModel):
    interaction_id: str
    reply_text: str

class OfflineReplyRequest(BaseModel):
    interaction_id: str
    reply_text: str


# ── 路由 ──

@router.get("/partner")
async def get_partner(
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """获取语伴信息"""
    result = await engine.get_partner(user_id)
    return {"status": "ok", "data": result}


@router.get("/greeting")
async def get_greeting(
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """获取一条问候语（模拟语伴发来的消息）"""
    result = await engine.get_greeting(user_id)
    return {"status": "ok", "data": result}


@router.post("/reply")
async def reply_greeting(
    body: ReplyRequest,
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """回复问候语 — AI 润色 + 模拟反馈 + 发放奖励"""
    try:
        result = await engine.reply_greeting(user_id, body.interaction_id, body.reply_text)
        return {"status": "ok", "data": result}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/history")
async def get_history(
    limit: int = 10,
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """获取桥梁交互历史"""
    result = await engine.get_history(user_id, limit)
    return {"status": "ok", "data": result}


@router.get("/progress")
async def get_progress(
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """获取桥梁进度"""
    result = await engine.get_progress(user_id)
    return {"status": "ok", "data": result}


# ── Lv.2 点赞桥 ──

@router.get("/like")
async def get_like_prompt(
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """获取一条点赞桥内容（语伴分享话题）"""
    result = await engine.get_like_prompt(user_id)
    return {"status": "ok", "data": result}


@router.post("/like")
async def like_reply(
    body: LikeRequest,
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """用户评价语伴分享的内容"""
    try:
        result = await engine.like_reply(user_id, body.interaction_id, body.reply_text)
        return {"status": "ok", "data": result}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


# ── Lv.3 梗桥 ──

@router.get("/teach")
async def get_teach_prompt(
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """获取一条教梗提示（语伴想学一个词）"""
    result = await engine.get_teach_prompt(user_id)
    return {"status": "ok", "data": result}


# ── Lv.4 群聊桥 ──

@router.get("/group")
async def get_group_prompt(
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """获取一个群聊场景"""
    result = await engine.get_group_prompt(user_id)
    return {"status": "ok", "data": result}


@router.post("/group")
async def group_reply(
    body: GroupReplyRequest,
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """用户在群聊中发言"""
    try:
        result = await engine.group_reply(user_id, body.interaction_id, body.reply_text)
        return {"status": "ok", "data": result}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


# ── Lv.5 线下桥 ──

@router.get("/offline")
async def get_offline_prompt(
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """获取一个线下场景"""
    result = await engine.get_offline_prompt(user_id)
    return {"status": "ok", "data": result}


@router.post("/offline")
async def offline_reply(
    body: OfflineReplyRequest,
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """用户在线下场景中回复"""
    try:
        result = await engine.offline_reply(user_id, body.interaction_id, body.reply_text)
        return {"status": "ok", "data": result}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/teach")
async def teach_word(
    body: TeachRequest,
    engine: BridgeEngine = Depends(get_bridge_engine),
    user_id: str = Depends(get_user_id),
):
    """用户教语伴一个词"""
    try:
        result = await engine.teach_word(user_id, body.interaction_id, body.word, body.meaning, body.example)
        return {"status": "ok", "data": result}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
